core/src/kmx/kmx_xstring.h

```diff
@@ -149,6 +171,50 @@ u16string_to_u32string(const std::u16string &source) {
   return out;
 }
 
+inline bool Uni_IsEndOfPlaneNonCharacter(km_kbp_usv ch) {
+  return (((ch) & Uni_FFFE_NONCHARACTER) == Uni_FFFE_NONCHARACTER); // matches FFFF or FFFE
+}
+
+inline bool Uni_IsNoncharacter(km_kbp_usv ch) {
+  return (((ch) >= Uni_FD_NONCHARACTER_START && (ch) <= Uni_FD_NONCHARACTER_END) || Uni_IsEndOfPlaneNonCharacter(ch));
+}
+
+inline bool Uni_InCodespace(km_kbp_usv ch) {
+  return ((ch) <= 0x10FFFF);
+};
+
+inline bool Uni_IsValid(km_kbp_usv ch) {
+  return (Uni_InCodespace(ch) && !Uni_IsSurrogate(ch) && !Uni_IsNoncharacter(ch));
+}
+
+inline bool Uni_IsValid(km_kbp_usv start, km_kbp_usv end) {
+  // quicker check
+  if (!Uni_IsValid(end) || !Uni_IsValid(start) || (end < start)) {  // start is checked below
+    return false;
+  }
+
+  // If 'start' is low enough in the BMP, we need to avoid (1) Surrogates and (2) the FDxx nonchars
+  if (start < Uni_FD_NONCHARACTER_END) {
```

Comment:
Do we need this first test? The child tests test start anyway?

Reply:
yes, it was meant as a heuristic but probably fails at it

Reply:
done